proxy.go

```diff
@@ -0,0 +1,292 @@
+package tunnel
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/koding/logging"
+	"github.com/koding/tunnel/proto"
+)
+
+// Proxy is responsible for forwarding remote connection to local server and writing the response back.
+type Proxy interface {
```

Comment:
Why not a `ProxyFunc` instead? I think it'd would be easier for a user to provide just a function, instead of creating wrapper types to just be able to implement an interface.

Why do we need an interface, do you think it may have more methods in future?